servant-server/test/Servant/ServerSpec.hs

```diff
@@ -684,6 +687,31 @@ responseHeadersSpec = describe "ResponseHeaders" $ do
         THW.request method "" [(hAccept, "crazy/mime")] ""
           `shouldRespondWith` 406
 
+-- }}}
+------------------------------------------------------------------------------
+-- * uverbResponseHeaderSpec {{{
+------------------------------------------------------------------------------
+type UVerbHeaderResponse = '[
+  WithStatus 200 (Headers '[Header "H1" Int] String),
+  WithStatus 404 String ]
+
+type UVerbResponseHeadersApi =
+       Capture "ok" Bool :> UVerb 'GET '[JSON] UVerbHeaderResponse
+
+uverbResponseHeadersServer :: Server UVerbResponseHeadersApi
+uverbResponseHeadersServer True = pure . Z . I . WithStatus $ addHeader 5 "foo"
+uverbResponseHeadersServer False = pure . S . Z . I . WithStatus $ "bar"
```

Comment:
I'm not sure, but:

```suggestion
uverbResponseHeadersServer True = respond . WithStatus $ addHeader 5 "foo"
uverbResponseHeadersServer False = respond . WithStatus $ "bar"
```

Reply:
That's very strange.  Why does `WithStatus k1 a1` not match both of the expected types?

Also, I think I remembered correctly, and [respond](https://hackage.haskell.org/package/servant-server-0.18.2/docs/Servant-Server-UVerb.html#v:respond) is indeed the idiom of choice here.  So something is wrong.  I probably botched the `UVerb` code that has been sitting on master already for a while in yet another interesting subtle way?

Reply:
It seems `respond` does not work very well when there are still metavariables left to be unified. I replaced `respond` with its definition `pure . inject . I` to get more meaningful error messages, and I got (among other errors):
```
servant-server      >     • Overlapping instances for Servant.API.UVerb.Union.UElem
servant-server      >                                   (WithStatus k1 a1)
servant-server      >                                   '[WithStatus 200 (Headers '[Header "H1" Int] String),
servant-server      >                                     WithStatus 404 String]
servant-server      >         arising from a use of ‘inject’
```
My understanding is that the type-class based mechanism for checking list membership cannot perform unification of the type variables involved.

One solution, besides manually lifting to the appropriate constructor with `S` and `Z` like now, is to fully annotate the argument to respond with its type:
```haskell
uverbResponseHeadersServer True = respond (WithStatus (addHeader 5 "foo") :: WithStatus 200 (Headers '[Header "H1" Int] String))
uverbResponseHeadersServer False = respond (WithStatus "bar" :: WithStatus 404 String)
```
Which one do you prefer?

Reply:
Ah, I remember now.  Yes, you figured out the problem and the solution.  I personally prefer `respond` with the type annotation, but I think you should make the call.

Reply:
Ah, I saw that you finished this, never mind.  I'll take another look.